gravitee-apim-rest-api/gravitee-apim-rest-api-management/gravitee-apim-rest-api-management-rest/src/main/java/io/gravitee/rest/api/management/rest/resource/EnvironmentAnalyticsResource.java

```diff
@@ -229,35 +216,51 @@ public String field() {
 
             query.setAggregations(aggregationList);
         }
-        addExtraFilter(query, extraFilter);
+        addExtraFilter(query, fieldFilter);
         return analyticsService.execute(executionContext, query);
     }
 
-    private Analytics executeGroupBy(final ExecutionContext executionContext, AnalyticsParam analyticsParam, String extraFilter) {
-        GroupByQuery query = new GroupByQuery();
-        query.setFrom(analyticsParam.getFrom());
-        query.setTo(analyticsParam.getTo());
-        query.setInterval(analyticsParam.getInterval());
-        query.setQuery(analyticsParam.getQuery());
-        query.setField(analyticsParam.getField());
-
-        if (analyticsParam.getOrder() != null) {
-            final GroupByQuery.Order order = new GroupByQuery.Order();
-            order.setField(analyticsParam.getOrder().getField());
-            order.setType(analyticsParam.getOrder().getType());
-            order.setOrder(analyticsParam.getOrder().isOrder());
-            query.setOrder(order);
+    /**
+     * @param executionContext
+     * @param analyticsParam
+     * @return
+     * @throws FieldFilterEmptyException: if user is not Admin and filter based on ids field is empty
+     */
+    private String buildFieldFilterForNonAdmin(ExecutionContext executionContext, AnalyticsParam analyticsParam)
+        throws FieldFilterEmptyException {
+        // add filter by Apis or Applications
+        if (isAdmin()) {
+            return null;
         }
 
-        List<Range> ranges = analyticsParam.getRanges();
-        if (ranges != null) {
-            Map<Double, Double> rangeMap = ranges.stream().collect(Collectors.toMap(Range::getFrom, Range::getTo));
-
-            query.setGroups(rangeMap);
+        String fieldName;
+        List<String> ids;
+        if (APPLICATION_FIELD.equalsIgnoreCase(analyticsParam.getField())) {
+            fieldName = APPLICATION_FIELD;
+            ids =
+                applicationService
+                    .findIdsByUser(executionContext, getAuthenticatedUser())
+                    .stream()
+                    .filter(appId -> permissionService.hasPermission(executionContext, APPLICATION_ANALYTICS, appId, READ))
+                    .collect(Collectors.toList());
+        } else {
+            fieldName = API_FIELD;
+            ids =
+                apiService
+                    .findIdsByUser(executionContext, getAuthenticatedUser(), null, false)
+                    .stream()
+                    .filter(apiId -> permissionService.hasPermission(executionContext, API_ANALYTICS, apiId, READ))
+                    .collect(Collectors.toList());
         }
 
-        addExtraFilter(query, extraFilter);
+        if (ids.isEmpty()) {
+            throw new FieldFilterEmptyException();
```

Comment:
@gaetanmaisse should we handle this case with an empty string instead of using Exception ?

Reply:
An `Optional<String>`? But I'm not convinced about the benefits vs readability

Reply:
yep, for me, it's more readable with an exception